google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java

```diff
@@ -1266,6 +1266,197 @@ public void testPostgreSQLGetQueryParameters() {
+    assertTrue(
+        parser
+            .parse(Statement.of("insert into x (a,b) values (1,2) then return(a)"))
+            .hasReturningClause());
+    assertTrue(
+        parser
+            .parse(Statement.of("insert into x (a,b) values (1,2)then return(a)"))
+            .hasReturningClause());
+    assertTrue(
+        parser
+            .parse(Statement.of("insert into x (a,b) values (1,2) then/*comment*/return *"))
+            .hasReturningClause());
+    assertTrue(
+        parser
+            .parse(Statement.of("insert into x (a,b) values (1,2) then return /*then return*/ *"))
+            .hasReturningClause());
+    assertTrue(
+        parser
+            .parse(Statement.of("insert into x (a,b) values (1,2)then/*comment*/return *"))
+            .hasReturningClause());
+    assertTrue(
+        parser
+            .parse(Statement.of("insert into x (a,b) values (1,2)then/*comment*/return(a)"))
+            .hasReturningClause());
+    assertTrue(
+        parser
+            .parse(Statement.of("insert into x (a,b) values (1,2)then /*comment*/return(a)"))
+            .hasReturningClause());
+    assertTrue(
+        parser
+            .parse(
+                Statement.of("insert into x (a,b) values (1,2)/*comment*/then/*comment*/return(a)"))
+            .hasReturningClause());
+    assertTrue(
+        parser
+            .parse(
+                Statement.of(
+                    "insert into x (a,b) values (1,2)/*comment*/then/*comment*/return/*comment*/(a)"))
+            .hasReturningClause());
+    assertTrue(
+        parser
+            .parse(
+                Statement.of(
+                    "insert into x (a,b) values (1,2)/*comment"
+                        + "*/then"
+                        + "/*comment"
+                        + "*/return/*"
+                        + "comment*/(a)"))
+            .hasReturningClause());
+    assertTrue(
+        parser
+            .parse(Statement.of("delete from x where y=\"z\"then return *"))
+            .hasReturningClause());
+    assertTrue(
+        parser
+            .parse(Statement.of("insert into x select 'then return' as returning then return *"))
+            .hasReturningClause());
+    assertTrue(
+        parser.parse(Statement.of("delete from x where 10=`z`then return *")).hasReturningClause());
+    assertFalse(
+        parser
+            .parse(Statement.of("insert into x (a,b) values (1,2) returning (a)"))
+            .hasReturningClause());
+    assertFalse(
+        parser
+            .parse(Statement.of("insert into x (a,b) values (1,2) /*then return **/"))
+            .hasReturningClause());
+    assertFalse(
+        parser.parse(Statement.of("insert into x (a,b) values (1,2)")).hasReturningClause());
+    assertFalse(
+        parser
+            .parse(Statement.of("insert into x (a,b) values (1,2)thenreturn*"))
+            .hasReturningClause());
+  }
+
+  @Test
+  public void testPostgreSQLReturningClause() {
```

Comment:
bvandiver=# insert into t1 select 2returning *;
 a 
---
 2
(1 row)

Reply:
Thanks for this test case. I've handled this scenario by checking if the substring containing returning doesn't start with a valid first character of an identifier. In that case, the substring can be divided further into substrings. Have added more tests around that as well.